motoman_driver/MotoPlus/IoServer.c

```diff
@@ -427,19 +453,184 @@ int Ros_IoServer_WriteIOGroup(SimpleMsg* receiveMsg, SimpleMsg* replyMsg)
 	replyMsg->header.msgType = ROS_MSG_MOTO_WRITE_IO_GROUP_REPLY;
 	replyMsg->header.commType = ROS_COMM_SERVICE_REPLY;
 
-	for (i = 0; i < 8; i += 1)
+	bAddressOk = Ros_IoServer_IsValidWriteAddress(receiveMsg->body.writeIOGroup.ioAddress, IO_ACCESS_GROUP);
+	bValueOk = Ros_IoServer_IsValidWriteValue(receiveMsg->body.writeIOGroup.ioValue, IO_ACCESS_GROUP);
+
+	if (bAddressOk && bValueOk)
+	{
+		for (i = 0; i < 8; i += 1)
+		{
+			ioWriteData[i].ulAddr = (receiveMsg->body.writeIOGroup.ioAddress * 10) + i;
+			ioWriteData[i].ulValue = (receiveMsg->body.writeIOGroup.ioValue & (1 << i)) >> i;
+		}
+		apiRet = mpWriteIO(ioWriteData, 8);
+
+		replyMsg->body.writeIOGroupReply.resultCode = (apiRet == OK) ? IO_RESULT_OK : IO_RESULT_WRITE_API_ERROR;
+		replyMsg->header.replyType = (apiRet == OK) ? ROS_REPLY_SUCCESS : ROS_REPLY_FAILURE;
+	}
+	else
 	{
-		ioWriteData[i].ulAddr = (receiveMsg->body.writeIOGroup.ioAddress * 10) + i;
-		ioWriteData[i].ulValue = (receiveMsg->body.writeIOGroup.ioValue & (1 << i)) >> i;
+		replyMsg->header.replyType = ROS_REPLY_FAILURE;
+
+		if (!bAddressOk)
+			replyMsg->body.writeIOGroupReply.resultCode = IO_RESULT_WRITE_ADDRESS_INVALID;
+		else if (!bValueOk)
+			replyMsg->body.writeIOGroupReply.resultCode = IO_RESULT_WRITE_VALUE_INVALID;
 	}
-	apiRet = mpWriteIO(ioWriteData, 8);
 
-	if (apiRet == OK)
-		resultCode = ROS_REPLY_SUCCESS;
+	return OK; //keep connection alive regardless of any error code
+}
+
+int Ros_IoServer_ReadIORegister(SimpleMsg* receiveMsg, SimpleMsg* replyMsg)
+{
+	int apiRet;
+	MP_IO_INFO ioReadInfo;
+	USHORT ioValue = 0;
+
+	//initialize memory
+	memset(replyMsg, 0x00, sizeof(SimpleMsg));
+
+	// set prefix: length of message excluding the prefix
+	replyMsg->prefix.length = sizeof(SmHeader) + sizeof(SmBodyMotoReadIOMRegisterReply);
+
+	// set header information of the reply
+	replyMsg->header.msgType = ROS_MSG_MOTO_READ_MREGISTER_REPLY;
+	replyMsg->header.commType = ROS_COMM_SERVICE_REPLY;
+
+	if (receiveMsg->body.readRegister.registerNumber < 1000000)
+		receiveMsg->body.readRegister.registerNumber += 1000000;
```

Comment:
Would it perhaps be better to check this and fail if `1e6` was added by the client?

If I understand the comments in `SimpleMessage.h` correctly, clients should specify the register index directly (ie: 1, 2, 3, etc), instead of adding `1e6` themselves.

For consistency, it may be better to reject `registerNumber > whatever_the_normal_max_is`.

Reply:
> If I understand the comments in SimpleMessage.h correctly, clients should specify the register index directly (ie: 1, 2, 3, etc), instead of adding 1e6 themselves.

Yes. But, I put this catch in there in case someone doesn't read the comment. 
I think it's better to allow for more flexible input format, rather than return an error code and force someone to waste time debugging just because they didn't read a comment in a header file. This extra check eliminates confusion and headache in the future.

Reply:
> I think it's better to allow for more flexible input format, rather than return an error code and force someone to waste time debugging just because they didn't read a comment in a header file. 

that's not the only option we have: we can create an appropriate result_code and return it to clients.

Reply:
Would returning an error result code be an option @ted-miller? Seems anything out of the range [0, 1000)` would be illegal.

Reply:
I don't see any reason to return an error. If the value is between 0 - 999 or 1000000 - 1000999, then we know what they're trying to do. We might as well honor the request message. There's no benefit to making it more difficult for the user.

Reply:
Reason I suggest returning an error is because it's very valuable to have interfaces which do one thing well, without possibility for ambiguity or allowing unneeded variation.

Allowing both `[0, 999]` and `[1000000, 1000999]` is such unneeded variation. Only by reading documentation would users know which ranges are accepted, and it would only be the documentation here iiuc, as the MotoPlus documentation actually only allows `[1000000, 1000999]`.

Seeing as that's a likely source of information for users / customers of Yaskawa robots, it would seem good to actually only accept `[1000000, 1000999]` and nothing else.

This would make it consistent with available documentation and the the rest of of the IO services we expose here.

Reply:
But I don't feel strongly enough about it to hold up this PR, so if you'd like this merged as-is now, we can do that.

Reply:
I've had too many occurrences of supporting a user over the course of multiple days, only to find that the root cause was a simple ID or value being misinterpreted. It rightfully upsets the user, and wastes both our time.

Anything that makes an interface more accepting so that it "just works" is a good thing. There is no downside to making something easier for different users. (Especially since this interface is being separated from the motoman-driver so that it can be used independently.)

Reply:
Ok. Let's agree to disagree. Personally I find clear error messages and clearly defined interfaces a good thing, especially if those interfaces adhere to what is documented (in documentation available elsewhere), but I can see how what you describe can be valuable.